accepted/2022/enhance-vector-codegen.md

```diff
@@ -0,0 +1,679 @@
+
+- The presence of an `if` sink/source will take precedence over the presence of `Vector` memory sinks, and will serve as the primary threshold check for sampling.
+
+- If no `if` sink/source is established, the default threshold check will be (2 * vector size). 
+
+
+For example, in the following method, the result of `span.Slice` is used as a argument to a `Vector` creation. The expression is marked as a sink:
+
+```C#
+#[Vectorize]
+public int SumVector(ReadOnlySpan<int> span) // <-- Source
+{
+  Vector<int> vresult = Vector<int>.Zero;
+
+  int lastBlockIndex = span.Length - (span.Length % Vector<int>.Count);
+  int i = 0;
+
+  for (i; i < lastBlockIndex; i += Vector<int>.Count)
+  {
+    vresult += new Vector<int>(span.Slice(i));  // <-- Sink
+  }
+
+  int result = vresult.Sum();
+
+  // Handle tail
+  while (i < span.Length)
+  {
+    result += span[(span.Length - rem)+i];
+  }
+
+  return result;
+}
+```
+
+will result in `span` being selected as a source for sampling. The JIT will _conceptually_ perform the following sampling and recompiling --- we elide implementation details for now, which we recognize must fit into the existing PGO architecture (no actual "stub" will exist, instead the recompile logic will take place inside RyuJIT during a Tier1 compile check etc. )
+
+```C#
+public int SumVector_Stub(ReadOnlySpan<int> span)
+{
+  int averageSpanLength = Sample(span.Length, spanLengthGlobal);
+  if (averageSpanLength >= 2 * Vector512<int>.Count )
+  {
+    // Recompile
+  }
+  // ...
+  SumVector(span);
+}
+```
+
+<br/>
+
+In the following sample, and `if` conditional contains a sink:
+
+```C#
+#[Vectorize]
+public static unsafe nuint NarrowUtf16ToAscii(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+{
+  // ...
+  nuint currentOffset = 0;
+  uint SizeOfVector = (uint)Unsafe.SizeOf<Vector<byte>>(); // JIT will make this a const
+
+  // Only bother vectorizing if we have enough data to do so.
+  if (elementCount >= 4 * SizeOfVector) // <-- Sink
+  {
+    // ...
+    
+  }
+}
+```
+
+which will result in `elementCount` being selected as a source, with a threshold defined based on the `if` conditional:
+
+```C#
+#[Vectorize]
+public static unsafe nuint NarrowUtf16ToAscii_Stub(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+{
+  int averageElementCount = Sample(elementCount, elementCountGlobalSample);
+  if (averageSpanLength >= 4 * Vector512<byte>.Count)
+  {
+    // Recompile
+  }
+  NarrowUtf16ToAscii(pUtf16Buffer, pAsciiBuffer, elementCount);
+}
+```
+#### 3. Codegen Design for Both
+
+##### Transitive Method Codegen
+
+As both proposals allow `Vector<T>` to be specialized per-method, we cannot simply pass `Vector<T>` as an argument to helper methods as before. To address this issue, we propose an additional `[Vectorizeable]` attribute which allows to JIT to specialize the method per selected vector width if used in a `Vectorize.If` or `[Vectorize]`. 
```

Comment:
This specialization would likely have to be driven by VM -- the JIT has no way of producing multiple method bodies per method.

This begs the question of how we would reconcile this with the current requirement that all Vector<T> within a given runtime instance are the same size.

Reply:
I'd rather expect `Vector<T>` to be lowered to Vector128, Vector256, Vector512 by some IL processing/SG  - it's much simpler to do. And yes, good point about variadic Vector<> - if it escapes current method or exposed as an argument then we can't do it

Reply:
I think these are fair points --- how to reconcile a PGO driven approach for Vector<T> with the current requirement that Vector<T> be a chosen size for the duration of a runtime instance is an open question.

@EgorBo part of the reason I came up with the `#[Vectorize]` attribute was for this very reason: If a method has `#[Vectorize]`, its use of `Vector<T>` is of a `Vector<T>` whose size is selected by PGO. Without `#[Vectorize]`, `Vector<T>` will codegen to the size determined by the runtime process as it currently does now.

I think @EgorBo suggestion about a runtime probe `RuntimeHelpers.MostylTrue` as a way to mitigate the need for a dataflow analysis for PGO, and the `#[Vectorize]` attribute can probably co-exist. The former drives where the probes happen, the latter indicates that this `Vector<T>` is selected per-method, not per-process.

Reply:
> variadic Vector<> - if it escapes current method or exposed as an argument then we can't do it

Noting that `VectorMask<T>` will have similar considerations. `SVE` for `Arm64` as well, where the technical description disallows its usage as a field and various other scenarios.

`Vector<T>` is our current "best fit" for SVE on Arm64 and it would be good if we can continue using it there and here to solve the Vector128/256/512<T> considerations for x64 as well.

If we can make it so that dealing with leading/trailing elements is efficient the exact size of the Vector isn't a huge concern as we no longer lose vectorization on "small data" when the hardware register size increases.

We then just need to consider the scenario where the "largest register" isn't the "best choice" for a given scenario and where using a smaller vector would be better. We could potentially handle this with block/loop cloning, special handling in the JIT/tiered compilation process, or one of several other ways. The biggest "pit of failure" then becomes how `Vector<T>` behaves when encountered as a field, pointer, or when passed/returned between methods.

In particular, SVE limits it accordingly:
> Because of their unknown size at compile time, SVE types must not be used:
* to declare or define a static or thread-local storage variable
* as the type of an array element
* as the operand to a new expression
* as the type of object deleted by a delete expression
* as the argument to sizeof and _Alignof
* with pointer arithmetic on pointers to SVE objects (this affects the +, -, ++, and -- operators)
* as members of unions, structures and classes
* in standard library containers like std::vector.

Naturally, due to back-compat `Vector<T>` violates most/all of these but some of them don't matter as much in the context of a `JIT` environment, only for an AOT environment. But, if we can resolve how to handle many of these such that the JIT can enable selecting `Vector128/256/512<T>` as the backing "per method" (with limitations), then we can likely use an analyzer and other functionality to help drive users toward success.